File: stacks/stack.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 10

typedef struct {
    int top;
    int stack_array[SIZE];
} *Stack, Stack_Type;

Stack init()
{
    Stack S = malloc(sizeof(Stack_Type));
    S->top = -1;
    return S;
}
int push(int elem, Stack S)
{
    if(S->top < SIZE-1) {
        ++(S->top);
        S->stack_array[S->top] = elem;
        return 1;
    } else
        printf("Stack full");
    return 0;
}

int pop(Stack S)
{
    if(S->top != -1){
        --(S->top);
        return S->stack_array[S->top];
    } else
        printf("Stack empty");
    return 0;
}

int peek(Stack S)
{
    if(S->top != -1)
        return S->stack_array[S->top];
    else
        printf("Stack empty");
    return 0;
}

void print_stack(Stack S)
{
    for(int i = 0; i <= S->top; i++)
        printf("%i\n", S->stack_array[i]);
}
int main(int argc, char **argv)
{
    Stack stack = init(stack);
    push(3, stack);
    push(5, stack);
    print_stack(stack);
    printf("popping...");
    pop(stack);
    print_stack(stack);
    push(8, stack);
    push(4, stack);
    printf("peek = %i\n\n", peek(stack));
    print_stack(stack); 
}

File: stacks/stack.py
#!/usr/bin/python

#base: bottom of stack, have been in stack the longest
#LIFO: last in, first out
#top: end of stack

class Stack(object):
    
    def __init__(self, size=10):
        
        self.size = size
        self._stack = [None] * size
        self.top = -1

    def push(self, elem):
        
        self.top += 1
        self._stack[self.top] = elem
        if self.top == self.size -1:
            self.size *= 2
            temp = self._stack
            self._stack = [None] * self.size
            for i in range(len(temp)):
                self._stack[i] = temp[i]

    def pop(self):
        
        if self.top == -1:
            print('Error: Popping from an empty stack')
            return None
        popped = self._stack[self.top]
        self._stack[self.top] = None
        self.top -= 1
        return popped
        
    def peek(self):

        return self._stack[self.top]

    def print_stack(self):

        print('Stack: ', end='')
        for e in self._stack:
            if e != None:
                print(e, end=' ')
    
    def get_size(self):

        stack_size = 0
        for e in self._stack:
            if e != None:
                stack_size += 1
        return stack_size
'''
s = Stack()
s.push(5)
for i in range(9):
    s.push(i)
s.print_stack()
print('\npeeking: {}'.format(s.peek()))
s.push(3)
s.print_stack()
print('\npeeking: {}'.format(s.peek()))
p = s.pop()
print('popped: {}'.format(p))
s.print_stack()
p = s.pop()
print('\npopped: {}'.format(p))
s.print_stack()
for i in range(1,9):
    s.push(i)
print('')
s.print_stack()
'''
